xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+    this.backoffPolicyProvider = checkNotNull(backoffPolicyProvider, "backoffPolicyProvider");
+    this.stopwatch = checkNotNull(stopwatch, "stopwatch");
+  }
+
+  @Override
+  void shutdown() {
+    channel.shutdown();
+    shutdownRpcStream();
+  }
+
+  @Override
+  void watchConfigData(String hostName, int port, ConfigWatcher watcher) {
+    checkState(configWatcher == null, "ConfigWatcher is already registered");
+    configWatcher = checkNotNull(watcher, "watcher");
+    this.hostName = checkNotNull(hostName, "hostName");
+    if (port == -1) {
+      ldsResourceName = hostName;
+    } else {
+      ldsResourceName = hostName + ":" + port;
+    }
+    if (rpcRetryTimer != null) {
+      // Currently in retry backoff.
+      return;
+    }
+    if (adsStream == null) {
+      startRpcStream();
+    }
+    adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName));
+  }
+
+  /**
+   * Builds a channel to the given server URI with the first supported channel creds config.
+   */
+  private static ManagedChannel buildChannel(String serverUri,List<ChannelCreds> channelCredsList) {
+    ManagedChannel ch = null;
+    // Use the first supported channel credentials configuration.
+    // Currently, only "google_default" is supported.
+    for (ChannelCreds creds : channelCredsList) {
+      if (creds.getType().equals("google_default")) {
+        ch = GoogleDefaultChannelBuilder.forTarget(serverUri).build();
+        break;
+      }
+    }
+    if (ch == null) {
+      ch = ManagedChannelBuilder.forTarget(serverUri).build();
+    }
+    return ch;
+  }
+
+  /**
+   * Establishes the RPC connection by creating a new RPC stream on the given channel for
+   * xDS protocol communication.
+   */
+  private void startRpcStream() {
+    checkState(adsStream == null, "Previous adsStream has not been cleared yet");
+    AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub =
+        AggregatedDiscoveryServiceGrpc.newStub(channel);
+    adsStream = new AdsStream(stub);
+    adsStream.start();
+    stopwatch.reset().start();
+  }
+
+  /**
+   * Terminates the RPC connection.
+   */
+  @VisibleForTesting
+  void shutdownRpcStream() {
+    if (adsStream != null) {
+      adsStream.close(Status.CANCELLED.withDescription("shutdown").asException());
+    }
+    if (rpcRetryTimer != null) {
+      rpcRetryTimer.cancel();
+    }
+  }
+
+  /**
+   * Handles LDS response to find the HttpConnectionManager message for the requested resource name.
+   * Proceed with the resolved RouteConfiguration in HttpConnectionManager message of the requested
+   * listener, if exists, to find the VirtualHost configuration for the "xds:" URI
+   * (with the port, if any, stripped off). Or sends an RDS request if configured for dynamic
+   * resolution. The response is NACKed if contains invalid data for gRPC's usage. Otherwise, an
+   * ACK request is sent to management server.
+   */
+  private void handleLdsResponse(DiscoveryResponse ldsResponse) {
```

Comment:
this function is huge: consider splitting it and based on my suggestion elsewhere consider using classes like `RouteConfigurationProcessor` to handle those respective data structures. Difficult to review/maintain/debug with the complexity in a single function.